router/core/router.go

```diff
@@ -795,6 +795,21 @@ func (r *Router) newServer(ctx context.Context, routerConfig *nodev1.RouterConfi
 	})
 	httpRouter.Use(recoveryHandler)
 	httpRouter.Use(middleware.RequestID)
+	// Rewrite request id in request header to allow "simple" forward to subgraph
```

Comment:
Hi @oxyno-zeta, this will work but we put such logic into [`preOriginHandlers`](https://github.com/wundergraph/cosmo/blob/a1005919a28ee8c9e9c3b87c730e47db336bceec/router/core/router.go#L258). This is the same way as custom module are developed and they are executed before the request to the subgraph. There is also a unit test example.

Reply:
Hello @StarpTech , are you saying that you want me to create a preOriginHandlers at this line ?

Like this ?

```golang
	r.preOriginHandlers = append(r.preOriginHandlers, func(req *http.Request, ctx RequestContext) (*http.Request, *http.Response) {
		// Get request id
		reqID := middleware.GetReqID(req.Context())

		// Save it in request header
		// Note: if it wasn't generated by middleware, it will override with the same value
		// Note: so nothing important is made in this case
		req.Header.Set(middleware.RequestIDHeader, reqID)

		// Next
		return req, nil
	})
	r.preOriginHandlers = append(r.preOriginHandlers, r.headerRuleEngine.OnOriginRequest)
```

Reply:
Take a look at this example. It implements the `EnginePreOriginHandler` interface. https://github.com/wundergraph/cosmo/blob/a1005919a28ee8c9e9c3b87c730e47db336bceec/router/core/header_rule_engine.go
Test: https://github.com/wundergraph/cosmo/blob/a1005919a28ee8c9e9c3b87c730e47db336bceec/router/core/header_rule_engine_test.go

Reply:
I'm very sorry @StarpTech . Maybe because it is late here but I don't understand what you looking for. Do you want to create a full header rule engine for this ?

Reply:
We must be in a similar timezone :smile: This

```go
package core

import (
	"github.com/wundergraph/cosmo/router/pkg/config"
	"net/http"
)

var (
	_ EnginePreOriginHandler = (*ForwardRequestHeader)(nil)
)

type ForwardRequestHeader struct {
}

func NewHeaderTransformer(rules config.HeaderRules) *ForwardRequestHeader {
	return &ForwardRequestHeader{}
}

func (h ForwardRequestHeader) OnOriginRequest(request *http.Request, ctx RequestContext) (*http.Request, *http.Response) {

	ctx.Request().Header.Set("X-Request-ID", request.Header.Get("X-Request-ID"))

	return request, nil
}

```

Then you can append  on `preOriginHandlers`

Reply:
Yes I think we are in the same one ;)

Oh I see what you are requesting. I will try that way !

Reply:
I've updated the code with your recommendation. This is working well.